File: plugins/plugin-trajectory-logger/rust/src/art_format.rs
use crate::types::{
    ARTTrajectory, ChatMessage, JsonValue, Trajectory, TrajectoryGroup, TrajectoryStep,
};
use std::collections::HashMap;

/// Convert a rich trajectory to an ART message array.
pub fn to_art_messages(trajectory: &Trajectory) -> Vec<ChatMessage> {
    let mut messages: Vec<ChatMessage> = Vec::new();

    if let Some(system) = build_system_message(trajectory) {
        messages.push(system);
    }

    for step in &trajectory.steps {
        if let Some(user) = build_user_message(step) {
            messages.push(ChatMessage {
                role: "user".to_string(),
                content: user,
                name: None,
            });
        }

        if let Some(assistant) = build_assistant_message(step) {
            messages.push(ChatMessage {
                role: "assistant".to_string(),
                content: assistant,
                name: None,
            });
        }
    }

    messages
}

fn build_system_message(trajectory: &Trajectory) -> Option<ChatMessage> {
    if let Some(first_step) = trajectory.steps.first() {
        if let Some(first_call) = first_step.llm_calls.first() {
            if !first_call.system_prompt.is_empty() {
                return Some(ChatMessage {
                    role: "system".to_string(),
                    content: first_call.system_prompt.clone(),
                    name: None,
                });
            }
        }
    }

    let agent_name = trajectory
        .metadata
        .get("agentName")
        .and_then(|v| v.as_str())
        .unwrap_or("Agent");
    let goal = trajectory
        .metadata
        .get("goalDescription")
        .and_then(|v| v.as_str())
        .unwrap_or("make good decisions");

    Some(ChatMessage {
        role: "system".to_string(),
        content: format!(
            "You are {}, an autonomous agent. Your goal is to {}.",
            agent_name, goal
        ),
        name: None,
    })
}

fn build_user_message(step: &TrajectoryStep) -> Option<String> {
    if let Some(call) = step
        .llm_calls
        .iter()
        .find(|c| matches!(c.purpose, crate::types::LLMPurpose::Action))
    {
        if !call.user_prompt.is_empty() {
            return Some(call.user_prompt.clone());
        }
    }

    let mut parts: Vec<String> = Vec::new();
    parts.push("Current state:".to_string());
    parts.push(format!(
        "- Balance: ${}",
        step.environment_state.agent_balance
    ));
    parts.push(format!("- P&L: ${}", step.environment_state.agent_pnl));
    parts.push(format!(
        "- Open Positions: {}",
        step.environment_state.open_positions
    ));

    for provider in &step.provider_accesses {
        parts.push(format!("\n{} data:", provider.provider_name));
        parts.push(
            serde_json::to_string_pretty(&provider.data)
                .unwrap_or_default()
                .to_string(),
        );
    }

    parts.push("\nWhat action should you take?".to_string());
    Some(parts.join("\n"))
}

fn build_assistant_message(step: &TrajectoryStep) -> Option<String> {
    if let Some(call) = step
        .llm_calls
        .iter()
        .find(|c| matches!(c.purpose, crate::types::LLMPurpose::Action))
    {
        if !call.response.is_empty() {
            return Some(call.response.clone());
        }
    }

    let action = &step.action;
    let mut parts: Vec<String> = vec![format!("I will {}.", action.action_type)];
    if let Some(reasoning) = &action.reasoning {
        parts.push(format!("Reasoning: {}", reasoning));
    }
    parts.push(format!(
        "Parameters: {}",
        serde_json::to_string(&action.parameters).unwrap_or_default()
    ));
    Some(parts.join("\n"))
}

/// Convert to ART trajectory object.
pub fn to_art_trajectory(trajectory: &Trajectory) -> ARTTrajectory {
    let mut metadata: HashMap<String, JsonValue> = HashMap::new();
    metadata.insert(
        "trajectoryId".to_string(),
        JsonValue::String(trajectory.trajectory_id.clone()),
    );
    metadata.insert(
        "agentId".to_string(),
        JsonValue::String(trajectory.agent_id.clone()),
    );
    if let Some(s) = &trajectory.scenario_id {
        metadata.insert("scenarioId".to_string(), JsonValue::String(s.clone()));
    }
    if let Some(g) = trajectory.group_index {
        metadata.insert(
            "groupIndex".to_string(),
            JsonValue::Number(serde_json::Number::from(g as u64)),
        );
    }
    metadata.insert(
        "metrics".to_string(),
        serde_json::to_value(&trajectory.metrics).unwrap_or(JsonValue::Null),
    );

    ARTTrajectory {
        messages: to_art_messages(trajectory),
        reward: trajectory.total_reward,
        metadata,
        metrics: Some(filter_numeric_metrics(&trajectory.metrics)),
    }
}

fn filter_numeric_metrics(metrics: &crate::types::TrajectoryMetrics) -> HashMap<String, f64> {
    let mut out: HashMap<String, f64> = HashMap::new();
    out.insert("episodeLength".to_string(), metrics.episode_length as f64);
    if let Some(v) = metrics.final_balance {
        out.insert("finalBalance".to_string(), v);
    }
    if let Some(v) = metrics.final_pnl {
        out.insert("finalPnL".to_string(), v);
    }
    if let Some(v) = metrics.success_rate {
        out.insert("successRate".to_string(), v);
    }
    out
}

/// Group trajectories by scenario id for GRPO.
pub fn group_trajectories(trajectories: &[Trajectory], now_ms: i64) -> Vec<TrajectoryGroup> {
    let mut map: HashMap<String, Vec<Trajectory>> = HashMap::new();
    for t in trajectories {
        let scenario = t
            .scenario_id
            .clone()
            .unwrap_or_else(|| "default".to_string());
        map.entry(scenario).or_default().push(t.clone());
    }

    map.into_iter()
        .enumerate()
        .map(|(idx, (scenario_id, trajs))| TrajectoryGroup {
            group_id: format!("group-{}", idx),
            scenario_id,
            shared_prefix: Some(extract_shared_prefix(&trajs)),
            trajectories: trajs,
            rankings: None,
            normalized_rewards: None,
            ruler_scores: None,
            created_at: now_ms,
            model_version: None,
        })
        .collect()
}

/// Extract shared prefix messages across trajectories.
pub fn extract_shared_prefix(trajectories: &[Trajectory]) -> Vec<ChatMessage> {
    if trajectories.is_empty() {
        return Vec::new();
    }

    let all_messages: Vec<Vec<ChatMessage>> = trajectories.iter().map(to_art_messages).collect();
    let first = match all_messages.first() {
        Some(m) => m,
        None => return Vec::new(),
    };

    let mut shared: Vec<ChatMessage> = Vec::new();
    for (i, msg) in first.iter().enumerate() {
        let all_match = all_messages.iter().all(|msgs| msgs.get(i) == Some(msg));
        if all_match {
            shared.push(msg.clone());
        } else {
            break;
        }
    }

    shared
}

/// Remove a shared prefix from messages.
pub fn remove_shared_prefix(
    messages: &[ChatMessage],
    shared_prefix: &[ChatMessage],
) -> Vec<ChatMessage> {
    messages.iter().skip(shared_prefix.len()).cloned().collect()
}

/// Prepare GRPO group for judge (shared prefix + per-trajectory suffixes).
pub fn prepare_for_ruler(
    group: &TrajectoryGroup,
) -> (
    Vec<ChatMessage>,
    Vec<Vec<ChatMessage>>,
    Vec<HashMap<String, JsonValue>>,
) {
    let shared = group
        .shared_prefix
        .clone()
        .unwrap_or_else(|| extract_shared_prefix(&group.trajectories));

    let art_trajs: Vec<ARTTrajectory> = group.trajectories.iter().map(to_art_trajectory).collect();
    let suffixes: Vec<Vec<ChatMessage>> = art_trajs
        .iter()
        .map(|t| remove_shared_prefix(&t.messages, &shared))
        .collect();

    let metadata: Vec<HashMap<String, JsonValue>> =
        art_trajs.into_iter().map(|t| t.metadata).collect();
    (shared, suffixes, metadata)
}

/// Validate trajectory can be converted to ART.
pub fn validate_art_compatibility(trajectory: &Trajectory) -> (bool, Vec<String>, Vec<String>) {
    let mut errors: Vec<String> = Vec::new();
    let mut warnings: Vec<String> = Vec::new();

    if trajectory.steps.is_empty() {
        errors.push("Trajectory has no steps".to_string());
    }

    for (idx, step) in trajectory.steps.iter().enumerate() {
        if step.llm_calls.is_empty() {
            errors.push(format!(
                "Step {} has no LLM calls - can't extract messages",
                idx
            ));
        }
        for call in &step.llm_calls {
            if call.user_prompt.len() < 10 {
                warnings.push(format!("Step {} has very short user prompt", idx));
            }
            if call.response.len() < 5 {
                warnings.push(format!("Step {} has very short response", idx));
            }
        }
    }

    let art = to_art_trajectory(trajectory);
    if art.messages.len() < 2 {
        warnings.push("Trajectory converts to very few messages (< 2)".to_string());
    }

    (errors.is_empty(), errors, warnings)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::{
        ActionAttempt, EnvironmentState, LLMCall, LLMPurpose, RewardComponents, Trajectory,
        TrajectoryMetrics,
    };
    use uuid::Uuid;

    #[test]
    fn test_to_art_messages() {
        let now = 1_700_000_000_000i64;
        let t = Trajectory {
            trajectory_id: Uuid::new_v4().to_string(),
            agent_id: Uuid::new_v4().to_string(),
            start_time: now,
            end_time: now,
            duration_ms: 0,
            episode_id: None,
            scenario_id: Some("s1".to_string()),
            batch_id: None,
            group_index: None,
            steps: vec![TrajectoryStep {
                step_id: Uuid::new_v4().to_string(),
                step_number: 0,
                timestamp: now,
                environment_state: EnvironmentState {
                    timestamp: now,
                    agent_balance: 100.0,
                    agent_points: 0.0,
                    agent_pnl: 0.0,
                    open_positions: 0,
                    active_markets: None,
                    portfolio_value: None,
                    unread_messages: None,
                    recent_engagement: None,
                    custom: None,
                },
                observation: HashMap::new(),
                llm_calls: vec![LLMCall {
                    call_id: Uuid::new_v4().to_string(),
                    timestamp: now,
                    model: "test-model".to_string(),
                    model_version: None,
                    system_prompt: "sys".to_string(),
                    user_prompt: "user prompt long enough".to_string(),
                    messages: None,
                    response: "assistant response".to_string(),
                    reasoning: None,
                    temperature: 0.7,
                    max_tokens: 32,
                    top_p: None,
                    prompt_tokens: None,
                    completion_tokens: None,
                    latency_ms: None,
                    purpose: LLMPurpose::Action,
                    action_type: None,
                }],
                provider_accesses: Vec::new(),
                reasoning: None,
                action: ActionAttempt {
                    attempt_id: Uuid::new_v4().to_string(),
                    timestamp: now,
                    action_type: "HOLD".to_string(),
                    action_name: "HOLD".to_string(),
                    parameters: HashMap::new(),
                    reasoning: None,
                    llm_call_id: None,
                    success: true,
                    result: None,
                    error: None,
                    immediate_reward: None,
                },
                reward: 0.5,
                done: true,
                metadata: None,
            }],
            total_reward: 0.5,
            reward_components: RewardComponents {
                environment_reward: 0.5,
                ..Default::default()
            },
            metrics: TrajectoryMetrics {
                episode_length: 1,
                ..Default::default()
            },
            metadata: HashMap::new(),
        };

        let messages = to_art_messages(&t);
        assert!(messages.len() >= 2);
    }
}

File: plugins/plugin-trajectory-logger/python/elizaos_plugin_trajectory_logger/art_format.py
from __future__ import annotations

from dataclasses import dataclass

from elizaos_plugin_trajectory_logger.types import (
    ARTTrajectory,
    ChatMessage,
    JsonValue,
    Trajectory,
    TrajectoryGroup,
    TrajectoryStep,
)


def to_art_messages(trajectory: Trajectory) -> list[ChatMessage]:
    messages: list[ChatMessage] = []

    system_message = _build_system_message(trajectory)
    if system_message is not None:
        messages.append(system_message)

    for step in trajectory.steps:
        user_content = _build_user_message(step)
        if user_content:
            messages.append(ChatMessage(role="user", content=user_content))

        assistant_content = _build_assistant_message(step)
        if assistant_content:
            messages.append(ChatMessage(role="assistant", content=assistant_content))

    return messages


def _build_system_message(trajectory: Trajectory) -> ChatMessage | None:
    if trajectory.steps and trajectory.steps[0].llm_calls:
        system_prompt = trajectory.steps[0].llm_calls[0].system_prompt
        if system_prompt:
            return ChatMessage(role="system", content=system_prompt)

    agent_name = str(trajectory.metadata.get("agentName") or "Agent")
    goal = str(trajectory.metadata.get("goalDescription") or "make good decisions")

    return ChatMessage(
        role="system",
        content=f"You are {agent_name}, an autonomous agent. Your goal is to {goal}.",
    )


def _build_user_message(step: TrajectoryStep) -> str | None:
    for call in step.llm_calls:
        if call.purpose == "action" and call.user_prompt:
            return call.user_prompt

    parts: list[str] = []
    parts.append("Current state:")
    parts.append(f"- Balance: ${step.environment_state.agent_balance}")
    parts.append(f"- P&L: ${step.environment_state.agent_pnl}")
    parts.append(f"- Open Positions: {step.environment_state.open_positions}")

    for provider in step.provider_accesses:
        parts.append(f"\n{provider.provider_name} data:")
        parts.append(str(provider.data))

    parts.append("\nWhat action should you take?")
    return "\n".join(parts)


def _build_assistant_message(step: TrajectoryStep) -> str | None:
    for call in step.llm_calls:
        if call.purpose == "action" and call.response:
            return call.response

    action = step.action
    parts: list[str] = [f"I will {action.action_type}."]
    if action.reasoning:
        parts.append(f"Reasoning: {action.reasoning}")
    parts.append(f"Parameters: {action.parameters}")
    return "\n".join(parts)


def to_art_trajectory(trajectory: Trajectory) -> ARTTrajectory:
    metadata: dict[str, JsonValue] = {
        "trajectoryId": trajectory.trajectory_id,
        "agentId": trajectory.agent_id,
        "scenarioId": trajectory.scenario_id,
        "groupIndex": trajectory.group_index,
        "metrics": trajectory.metrics.model_dump(mode="json"),
        # Preserve any caller-provided metadata for training/analysis.
        # We keep it namespaced to avoid colliding with reserved ART keys above.
        "extra": trajectory.metadata,
    }

    return ARTTrajectory(
        messages=to_art_messages(trajectory),
        reward=trajectory.total_reward,
        metadata=metadata,
        metrics=_filter_numeric_metrics(trajectory.metrics.model_dump(mode="json")),
    )


def _filter_numeric_metrics(metrics: dict[str, JsonValue]) -> dict[str, float]:
    numeric: dict[str, float] = {}
    for k, v in metrics.items():
        if isinstance(v, (int, float)):
            numeric[k] = float(v)
    return numeric


def group_trajectories(trajectories: list[Trajectory]) -> list[TrajectoryGroup]:
    groups: dict[str, list[Trajectory]] = {}
    for t in trajectories:
        scenario = t.scenario_id or "default"
        groups.setdefault(scenario, []).append(t)

    out: list[TrajectoryGroup] = []
    for idx, (scenario_id, trajs) in enumerate(groups.items()):
        out.append(
            TrajectoryGroup(
                group_id=f"group-{idx}",
                scenario_id=scenario_id,
                trajectories=trajs,
                shared_prefix=extract_shared_prefix(trajs),
                created_at=_now_ms(),
            )
        )
    return out


def extract_shared_prefix(trajectories: list[Trajectory]) -> list[ChatMessage]:
    if not trajectories:
        return []

    all_messages = [to_art_messages(t) for t in trajectories]
    if not all_messages or not all_messages[0]:
        return []

    first = all_messages[0]
    shared: list[ChatMessage] = []

    for i, message in enumerate(first):
        if all(i < len(msgs) and msgs[i] == message for msgs in all_messages):
            shared.append(message)
        else:
            break

    return shared


def remove_shared_prefix(
    messages: list[ChatMessage], shared_prefix: list[ChatMessage]
) -> list[ChatMessage]:
    return messages[len(shared_prefix) :]


@dataclass(frozen=True)
class RulerPayload:
    shared_prefix: list[ChatMessage]
    suffixes: list[list[ChatMessage]]
    metadata: list[dict[str, JsonValue]]


def prepare_for_ruler(group: TrajectoryGroup) -> RulerPayload:
    art_trajs = [to_art_trajectory(t) for t in group.trajectories]
    shared = group.shared_prefix or extract_shared_prefix(group.trajectories)

    return RulerPayload(
        shared_prefix=shared,
        suffixes=[remove_shared_prefix(t.messages, shared) for t in art_trajs],
        metadata=[t.metadata for t in art_trajs],
    )


def validate_art_compatibility(trajectory: Trajectory) -> tuple[bool, list[str], list[str]]:
    errors: list[str] = []
    warnings: list[str] = []

    if not trajectory.steps:
        errors.append("Trajectory has no steps")

    for idx, step in enumerate(trajectory.steps):
        if not step.llm_calls:
            errors.append(f"Step {idx} has no LLM calls - can't extract messages")
        for call in step.llm_calls:
            if not call.user_prompt or len(call.user_prompt) < 10:
                warnings.append(f"Step {idx} has very short user prompt")
            if not call.response or len(call.response) < 5:
                warnings.append(f"Step {idx} has very short response")

    if trajectory.total_reward is None:
        errors.append("Trajectory has no valid reward")

    art = to_art_trajectory(trajectory)
    if len(art.messages) < 2:
        warnings.append("Trajectory converts to very few messages (< 2)")

    return (len(errors) == 0, errors, warnings)


def _now_ms() -> int:
    return int(__import__("time").time() * 1000)
